server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/QueryParser.java

```diff
@@ -0,0 +1,172 @@
+import static org.apache.james.webadmin.vault.routes.QueryParser.Operator.equals;
+import static org.apache.james.webadmin.vault.routes.QueryParser.Operator.equalsIgnoreCase;
+
+import java.time.ZonedDateTime;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import javax.mail.internet.AddressException;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.vault.search.Criterion;
+import org.apache.james.vault.search.CriterionFactory;
+import org.apache.james.vault.search.Query;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableTable;
+
+class QueryParser {
+
+    static class QueryParserException extends Exception {
+        QueryParserException(String message) {
+            super(message);
+        }
+    }
+
+    enum FieldName {
+        deletionDate,
+        deliveryDate,
+        recipients,
+        sender,
+        hasAttachment,
+        originMailboxes,
+        subject;
+
+        static FieldName getField(String fieldNameString) throws QueryParserException {
+            return Stream.of(values())
+                .filter(fieldName -> fieldName.name().equals(fieldNameString))
+                .findFirst()
+                .orElseThrow(Throwing.supplier(() -> new QueryParserException("fieldName: '" + fieldNameString + "' is not supported")));
+        }
+    }
+
+    enum Operator {
+        equals,
+        equalsIgnoreCase,
+        contains,
+        containsIgnoreCase,
+        beforeOrEquals,
+        afterOrEquals;
+
+        static Operator getOperator(String operator) throws QueryParserException {
+            return Stream.of(values())
+                .filter(op -> op.name().equals(operator))
+                .findFirst()
+                .orElseThrow(() -> new QueryParserException("operator: '" + operator + "' is not supported"));
+        }
+    }
+
+    interface FieldParser<T> {
+
+        class MailboxIdParser implements FieldParser<MailboxId> {
+
+            final MailboxId.Factory mailboxIdFactory;
+
+            MailboxIdParser(MailboxId.Factory mailboxIdFactory) {
+                this.mailboxIdFactory = mailboxIdFactory;
+            }
+
+            @Override
+            public MailboxId parse(String mailboxIdString) {
+                return mailboxIdFactory.fromString(mailboxIdString);
+            }
+        }
+
+        FieldParser<ZonedDateTime> ZONED_DATE_TIME_PARSER = ZonedDateTime::parse;
+        FieldParser<String> STRING_PARSER = input -> input;
+        FieldParser<Boolean> BOOLEAN_PARSER = Boolean::valueOf;
+        FieldParser<MailAddress> MAIL_ADDRESS_PARSER = mailAddressString -> {
+            return Throwing.supplier(() -> parseMailAddress(mailAddressString)).sneakyThrow().get();
```

Comment:
The use of Throwing supplier here seems abusive. How about having a try/catch explicitly re-throwing an IllegalArgumentException?